File: 2019/day02/day02.ts
import { readFileSync } from 'fs';

function exec_one(opcode: number, v1: number, v2: number) {
  return opcode == 1 ? v1 + v2 : v1 * v2;
}

function part1(arr: number[], v1: number = 12, v2: number = 2) {
  arr = arr.map(v => v);
  arr[1] = v1;
  arr[2] = v2;
  for (let i = 0; i < arr.length && arr[i] != 99; i += 4)
    arr[arr[i + 3]] = exec_one(arr[i], arr[arr[i + 1]], arr[arr[i + 2]]);
  return arr[0];
}

function part2(vals: number[]) {
  for (let v1 = 0; v1 < 100; v1++)
    for (let v2 = 0; v2 < 100; v2++)
      if (part1(vals, v1, v2) == 19690720)
        return v1 * 100 + v2;
}

let data = readFileSync('input', 'utf-8');
let vals = data.trim().split(',').map(v => parseInt(v));

const ans1 = part1(vals);
console.log(`part1: ${ans1}`);

const ans2 = part2(vals);
console.log(`part2: ${ans2}`);

File: 2019/day02/day02.py
#!/usr/bin/env python
import sys


def run_simulation(vals, val1, val2):
    vals[1] = val1
    vals[2] = val2
    try:
        pc = 0
        while pc < len(vals):
            opcode = vals[pc]
            ind1, ind2, ind3 = vals[pc + 1:pc + 4]
            if opcode == 1:
                vals[ind3] = vals[ind1] + vals[ind2]
            elif opcode == 2:
                vals[ind3] = vals[ind1] * vals[ind2]
            elif opcode == 99:
                break
            pc += 4
    except IndexError:
        pass

    return vals[0]


vals = list(map(int, sys.stdin.read().split(',')))

part1 = run_simulation(vals.copy(), 12, 2)
print('part1:', part1)

part2 = None
for noun in range(100):
    for verb in range(100):
        res = run_simulation(vals.copy(), noun, verb)
        if res == 19690720:
            part2 = 100 * noun + verb

print('part2:', part2)

assert part1 == 6568671
assert part2 == 3951
